shortfin/python/shortfin_apps/llm/components/kvcache/trie_attention_cache.py

```diff
@@ -271,101 +298,110 @@ def allocate(
-                    f"TriePagedAttentionCache: Lookup found {len(matched_pages)} cached pages for token length {len(tokens)}"
-                )
+            if not cache_info:
+                raise ValueError("cache_info cannot be None")
+
+            cur_node = cache_info.last_cached_node
 
-                cached_pages = matched_pages
-                n_cached_tokens = 0
-                if matched_pages:
-                    n_cached_tokens = len(matched_pages) * self.tokens_per_page
-                remaining_length = len(tokens) - n_cached_tokens
-                n_empty_pages = math.ceil(remaining_length / self.tokens_per_page)
-            else:
-                n_empty_pages = math.ceil(len(tokens) / self.tokens_per_page)
+            n_empty_pages = math.ceil(len(tokens) / self.tokens_per_page)
 
-            if not cached_pages and allocation_block_size > 0:
-                n_empty_pages = allocation_block_size
+            if allocation_block_size > 0:
+                n_empty_pages = max(n_empty_pages, allocation_block_size)
 
             new_pages = self.page_pool.acquire_free_pages(n_empty_pages)
 
             if new_pages is None and evict:
                 # Try eviction
-                self.evict_pages(n_empty_pages - len(self.page_pool.available_pages))
+                number_evicted_pages = self.evict_pages(
+                    n_empty_pages - len(self.page_pool.available_pages)
+                )
                 new_pages = self.page_pool.acquire_free_pages(n_empty_pages)
 
                 if new_pages is None:
                     raise CacheAllocationFailure(
                         "Failed to acquire pages even after attempting eviction from LRU leaves"
                     )
 
-            cur_node.ref_count.increment()
-            pages = cached_pages + new_pages
-            self._allocated_pages.extend(new_pages)
+            if new_pages is None:
+                raise CacheAllocationFailure(
+                    "Failed to acquire pages and eviction is disabled"
+                )
 
-            num_tokens = len(tokens)
-            if cache_info:
-                if (
-                    cache_info.last_cached_node
-                    and not cache_info.last_cached_node.ref_count.is_empty()
+            if len(new_pages) > 0:
+                # some new pages are allocated and will be used to create children of cur_node, hence increment ref_count of cur_node.
+                # do not increment last_cached_node ref_count when we allocate along the same branch again
+                if len(cache_info.pages) == 0 or (
+                    len(cache_info.pages) > 0
+                    and cur_node.page.index == cache_info.pages[-1].index
                 ):
-                    cache_info.last_cached_node.ref_count.decrement()
-                pages = cache_info.pages + pages
-                num_tokens += cache_info.num_tokens
+                    cur_node.ref_count.increment()
```

Comment:
Directly manipulating the ref count for the node seems concerning. Is there a higher level idea we can do here to avoid managing ref counts in multiple locations?